File: bot.py
"""
Telegram Bot - Shopee Video Downloader
Bot que recebe links da Shopee e envia o vídeo sem marca d'água
"""

import os
import telebot
from telebot import types
from shopee_service import ShopeeDownloader
from database import UserManager
from payment_service import PaymentService

# ============================================================
# CONFIGURAÇÃO - Substitua pelo seu token do BotFather
# ============================================================
BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "SEU_TOKEN_AQUI")

# Inicializa componentes
bot = telebot.TeleBot(BOT_TOKEN)
downloader = ShopeeDownloader(output_dir="output_video")
user_manager = UserManager()
payment_service = PaymentService()

def send_payment_options(message, reason="limit_reached"):
    """
    Envia as opções de pagamento (Pix) e botões.
    """
    chat_id = message.chat.id
    user_id = message.from_user.id
    
    # Mensagem inicial
    msg_header = ""
    if reason == "limit_reached":
        msg_header = "🚫 *O seu limite diário gratuito acabou!*"
    elif reason == "expired":
        msg_header = "⚠️ *O seu plano Premium venceu!*"
    else:
        msg_header = "💎 *Plano Premium (Ilimitado)*"

    bot.send_message(
        chat_id,
        f"{msg_header}\n\n"
        "Renove agora para continuar baixando vídeos ilimitadamente e sem filas!",
        parse_mode="Markdown"
    )

    # Gera Pagamento Pix
    payment_data = payment_service.create_pix_payment(user_id)
    payment_id = payment_data['payment_id']
    pix_code = payment_data['pix_copy_paste']

    # Mensagem com Código Pix
    pix_msg = (
        "Seu código pix (Ilimitado - 30 Dias)\n"
        "Clique abaixo para copiar:\n\n"
        f"`{pix_code}`\n\n"
        "⚠️ _A PUSHIN PAY atua apenas como processadora de pagamentos._"
    )

    # Botões
    markup = types.InlineKeyboardMarkup(row_width=1)
    btn_check = types.InlineKeyboardButton("✅ Verificar Pagamento", callback_data=f"check_pay_{payment_id}")
    btn_support = types.InlineKeyboardButton("💬 Suporte", url="https://t.me/seusuporte") # Placeholder
    markup.add(btn_check, btn_support)

    bot.send_message(chat_id, pix_msg, parse_mode="Markdown", reply_markup=markup)


@bot.message_handler(commands=['start', 'help'])
def send_welcome(message):
    """Responde ao comando /start e /help."""
    welcome_text = """
🛒 *Shopee Video Downloader Bot*

Envie um link de vídeo da Shopee e eu baixo para você sem marca d'água!

*Comandos:*
/plano - Ver status do plano ou assinar Premium
/ilimitado - Assinar plano Ilimitado (Premium)

*Links suportados:*
• shopee.com.br
• shp.ee
• sv.shopee.com.br

_Bot desenvolvido para fins educacionais._
    """
    bot.reply_to(message, welcome_text, parse_mode="Markdown")


@bot.message_handler(commands=['plano', 'ilimitado'])
def handle_plan_command(message):
    """Responde aos comandos de plano."""
    user_id = message.from_user.id
    status = user_manager.check_allowance(user_id)
    
    if status['is_premium']:
        bot.reply_to(
            message,
            "💎 *Você é PREMIUM!*\n\n"
            "Seu plano é ilimitado.\n"
            "Aproveite!",
            parse_mode="Markdown"
        )
    else:
        send_payment_options(message, reason="command")


@bot.message_handler(func=lambda m: downloader.is_shopee_url(m.text or ""))
def handle_shopee_link(message):
    """Processa links da Shopee com verificação de cota."""
    chat_id = message.chat.id
    user_id = message.from_user.id
    shopee_url = message.text.strip()
    
    # 1. Verifica cota do usuário
    status = user_manager.check_allowance(user_id)
    
    if not status['allowed']:
        # Limite atingido
        send_payment_options(message, reason="limit_reached")
        return

    # Envia ação de "enviando vídeo"
    bot.send_chat_action(chat_id, 'upload_video')
    
    # Envia mensagem de processamento
    status_msg = bot.reply_to(message, "⏳ Baixando o vídeo... Aguarde!")
    
    filepath = None
    try:
        # Faz o download
        filepath = downloader.download(shopee_url)
        
        # Incrementa uso após sucesso
        user_manager.increment_usage(user_id)
        
        # Prepara texto de rodapé
        if status['is_premium']:
            footer_text = "💎 Usuário Premium (Ilimitado)"
        else:
            # Recalcula para mostrar o valor correto pós-incremento
            # (Poderíamos otimizar, mas assim é seguro)
            new_status = user_manager.check_allowance(user_id)
            remaining = new_status['downloads_left']
            footer_text = f"📉 Downloads restantes hoje: {remaining}/{user_manager.DAILY_LIMIT}"

        # Envia o vídeo
        bot.send_chat_action(chat_id, 'upload_video')
        with open(filepath, 'rb') as video_file:
            bot.send_video(
                chat_id, 
                video_file,
                caption=f"✅ Vídeo da Shopee sem marca d'água!\n\n_{footer_text}_\n\n_Lembre-se de creditar o criador original._",
                parse_mode="Markdown",
                supports_streaming=True
            )
        
        # Deleta a mensagem de status
        bot.delete_message(chat_id, status_msg.message_id)
        
    except Exception as e:
        # Em caso de erro, notifica o usuário
        bot.edit_message_text(
            f"❌ Erro ao baixar o vídeo:\n`{str(e)}`",
            chat_id,
            status_msg.message_id,
            parse_mode="Markdown"
        )
        
    finally:
        # IMPORTANTE: Limpa o arquivo do disco
        if filepath and os.path.exists(filepath):
            try:
                os.remove(filepath)
            except OSError:
                pass


@bot.callback_query_handler(func=lambda call: call.data.startswith("check_pay_"))
def callback_check_payment(call):
    """Verifica o status do pagamento."""
    payment_id = call.data.replace("check_pay_", "")
    user_id = call.from_user.id
    
    # Verifica status (Mock sempre retorna True)
    paid = payment_service.check_payment_status(payment_id)
    
    if paid:
        # Adiciona 30 dias de premium
        user_manager.add_premium_time(user_id, 30)
        
        # Edita a mensagem para confirmar sucesso
        try:
            bot.edit_message_text(
                "✅ *Pagamento confirmado!*\n\n"
                "Você agora é **Premium** por 30 dias.\n"
                "Downloads ilimitados liberados! 🚀",
                call.message.chat.id,
                call.message.message_id,
                parse_mode="Markdown"
            )
            bot.answer_callback_query(call.id, "Pagamento Aprovado!")
        except Exception:
            # Caso não consiga editar a mensagem
            bot.send_message(call.message.chat.id, "✅ Pagamento confirmado! Você agora é Premium.")
    else:
        bot.answer_callback_query(call.id, "Pagamento ainda não confirmado. Tente novamente em instantes.", show_alert=True)


@bot.callback_query_handler(func=lambda call: call.data == "buy_premium")
def callback_buy_premium_legacy(call):
    """Mapeia botão antigo para novo fluxo."""
    bot.answer_callback_query(call.id)
    send_payment_options(call.message, reason="command")


@bot.message_handler(func=lambda m: True)
def handle_other_messages(message):
    """Responde a mensagens desconhecidas."""
    bot.reply_to(
        message, 
        "🤔 *Comando não reconhecido.*\n\n"
        "Envie um link da Shopee para baixar ou use /plano para ver opções Premium.",
        parse_mode="Markdown"
    )


if __name__ == "__main__":
    print("🤖 Bot iniciado (v2.0 - Payment Ready)! Pressione Ctrl+C para parar.")
    print(f"📁 Output dir: {os.path.abspath(downloader.output_dir)}")
    print(f"💾 Database: {user_manager.DB_NAME}")
    
    # Remove webhook e inicia polling
    bot.remove_webhook()
    bot.infinity_polling(timeout=60, long_polling_timeout=60)

File: bot.js
/**
 * Telegram Bot - Shopee Video Downloader
 * Bot que recebe links da Shopee e envia o vídeo sem marca d'água
 */

const TelegramBot = require('node-telegram-bot-api');
const fs = require('fs');
const path = require('path');

const ShopeeDownloader = require('./shopeeService');
const UserManager = require('./database');
const PaymentService = require('./paymentService');

// ============================================================
// CONFIGURAÇÃO
// ============================================================
const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;

if (!BOT_TOKEN || BOT_TOKEN === 'SEU_TOKEN_AQUI') {
    console.error('❌ Erro: A variável de ambiente TELEGRAM_BOT_TOKEN não foi configurada!');
    process.exit(1);
}

// Inicializa componentes
const bot = new TelegramBot(BOT_TOKEN, { polling: true });
const downloader = new ShopeeDownloader('output_video');
const userManager = new UserManager();
const paymentService = new PaymentService();

// ============================================================
// INICIALIZAÇÃO ASSÍNCRONA
// ============================================================
async function init() {
    await userManager.init();
    console.log('🤖 Bot iniciado (v2.0 - Node.js)!');
    console.log(`📁 Output dir: ${path.resolve(downloader.outputDir)}`);
    console.log(`💾 Database: ${userManager.dbPath}`);
}

init().catch(err => {
    console.error('Erro na inicialização:', err);
    process.exit(1);
});

// ============================================================
// FUNÇÕES AUXILIARES
// ============================================================

async function sendPaymentOptions(chatId, userId, reason = 'limit_reached') {
    let msgHeader = '';
    if (reason === 'limit_reached') {
        msgHeader = '🚫 *O seu limite diário gratuito acabou!*';
    } else if (reason === 'expired') {
        msgHeader = '⚠️ *O seu plano Premium venceu!*';
    } else {
        msgHeader = '💎 *Plano Premium (Ilimitado)*';
    }

    await bot.sendMessage(chatId,
        `${msgHeader}\n\nRenove agora para continuar baixando vídeos ilimitadamente e sem filas!`,
        { parse_mode: 'Markdown' }
    );

    // Gera Pagamento Pix
    const paymentData = await paymentService.createPixPayment(userId);
    const paymentId = paymentData.payment_id;
    const pixCode = paymentData.pix_copy_paste;

    const pixMsg =
        `Seu código pix (Ilimitado - 30 Dias)\n` +
        `Clique abaixo para copiar:\n\n` +
        `\`${pixCode}\`\n\n` +
        `⚠️ _A PUSHIN PAY atua apenas como processadora de pagamentos._`;

    const keyboard = {
        inline_keyboard: [
            [{ text: '✅ Verificar Pagamento', callback_data: `check_pay_${paymentId}` }],
            [{ text: '💬 Suporte', url: 'https://t.me/seusuporte' }]
        ]
    };

    await bot.sendMessage(chatId, pixMsg, {
        parse_mode: 'Markdown',
        reply_markup: keyboard
    });
}

// ============================================================
// HANDLERS
// ============================================================

// Comando /start e /help
bot.onText(/\/(start|help)/, (msg) => {
    const welcomeText = `
🛒 *Shopee Video Downloader Bot*

Envie um link de vídeo da Shopee e eu baixo para você sem marca d'água!

*Comandos:*
/plano - Ver status do plano ou assinar Premium
/ilimitado - Assinar plano Ilimitado (Premium)

*Links suportados:*
• shopee.com.br
• shp.ee
• sv.shopee.com.br

_Bot desenvolvido para fins educacionais._
    `;
    bot.sendMessage(msg.chat.id, welcomeText, { parse_mode: 'Markdown' });
});

// Comando /plano e /ilimitado
bot.onText(/\/(plano|ilimitado)/, (msg) => {
    const userId = msg.from.id;
    const chatId = msg.chat.id;
    const status = userManager.checkAllowance(userId);

    if (status.is_premium) {
        bot.sendMessage(chatId,
            '💎 *Você é PREMIUM!*\n\nSeu plano é ilimitado.\nAproveite!',
            { parse_mode: 'Markdown' }
        );
    } else {
        sendPaymentOptions(chatId, userId, 'command');
    }
});

// Handler para links da Shopee
bot.on('message', async (msg) => {
    // Ignora comandos
    if (msg.text && msg.text.startsWith('/')) return;

    const text = msg.text || '';
    const chatId = msg.chat.id;
    const userId = msg.from.id;

    // Verifica se é um link da Shopee
    if (!ShopeeDownloader.isShopeeUrl(text)) {
        bot.sendMessage(chatId,
            '🤔 *Comando não reconhecido.*\n\nEnvie um link da Shopee para baixar ou use /plano para ver opções Premium.',
            { parse_mode: 'Markdown' }
        );
        return;
    }

    const shopeeUrl = text.trim();

    // 1. Verifica cota do usuário
    const status = userManager.checkAllowance(userId);

    if (!status.allowed) {
        await sendPaymentOptions(chatId, userId, 'limit_reached');
        return;
    }

    // Envia ação de "enviando vídeo"
    bot.sendChatAction(chatId, 'upload_video');

    // Envia mensagem de processamento
    const statusMsg = await bot.sendMessage(chatId, '⏳ Baixando o vídeo... Aguarde!');

    let filepath = null;
    try {
        // Faz o download
        filepath = await downloader.download(shopeeUrl);

        // Incrementa uso após sucesso
        userManager.incrementUsage(userId);

        // Prepara texto de rodapé
        let footerText;
        if (status.is_premium) {
            footerText = '💎 Usuário Premium (Ilimitado)';
        } else {
            const newStatus = userManager.checkAllowance(userId);
            footerText = `📉 Downloads restantes hoje: ${newStatus.downloads_left}/${userManager.DAILY_LIMIT}`;
        }

        // Envia o vídeo
        bot.sendChatAction(chatId, 'upload_video');
        await bot.sendVideo(chatId, filepath, {
            caption: `✅ Vídeo da Shopee sem marca d'água!\n\n_${footerText}_\n\n_Lembre-se de creditar o criador original._`,
            parse_mode: 'Markdown'
        });

        // Deleta a mensagem de status
        bot.deleteMessage(chatId, statusMsg.message_id).catch(() => { });

    } catch (error) {
        console.error('Erro no download:', error);
        bot.editMessageText(
            `❌ Erro ao baixar o vídeo:\n\`${error.message}\``,
            {
                chat_id: chatId,
                message_id: statusMsg.message_id,
                parse_mode: 'Markdown'
            }
        ).catch(() => { });

    } finally {
        // IMPORTANTE: Limpa o arquivo do disco
        if (filepath && fs.existsSync(filepath)) {
            try {
                fs.unlinkSync(filepath);
            } catch (e) {
                console.error('Erro ao deletar arquivo:', e);
            }
        }
    }
});

// Callback para verificar pagamento
bot.on('callback_query', async (query) => {
    const data = query.data;
    const chatId = query.message.chat.id;
    const messageId = query.message.message_id;
    const userId = query.from.id;

    if (data.startsWith('check_pay_')) {
        const paymentId = data.replace('check_pay_', '');
        const paid = await paymentService.checkPaymentStatus(paymentId);

        if (paid) {
            // Adiciona 30 dias de premium
            userManager.addPremiumTime(userId, 30);

            try {
                await bot.editMessageText(
                    '✅ *Pagamento confirmado!*\n\nVocê agora é **Premium** por 30 dias.\nDownloads ilimitados liberados! 🚀',
                    { chat_id: chatId, message_id: messageId, parse_mode: 'Markdown' }
                );
                bot.answerCallbackQuery(query.id, { text: 'Pagamento Aprovado!' });
            } catch (e) {
                bot.sendMessage(chatId, '✅ Pagamento confirmado! Você agora é Premium.');
            }
        } else {
            bot.answerCallbackQuery(query.id, {
                text: 'Pagamento ainda não confirmado. Tente novamente em instantes.',
                show_alert: true
            });
        }
    } else if (data === 'buy_premium') {
        bot.answerCallbackQuery(query.id);
        await sendPaymentOptions(chatId, userId, 'command');
    }
});

// Tratamento de erros
bot.on('polling_error', (error) => {
    console.error('Polling error:', error.code, error.message);
});

process.on('SIGINT', () => {
    console.log('\n🛑 Encerrando bot...');
    userManager.close();
    process.exit(0);
});
